Implement Robust Toolbar Support

dapp opened the pull request:
This commit seeks to improve toolbar support by
expanding the ability of ScreenPresenters to
present the toolbar

Additionally, improvements have been made to
simplify component event infrastructure by
making component event handling something
Screen presenters can do by default, instead
of having specific SP subclasses that do that.

Non-event handling screen presenters are still
available via an added Default implementation.

fkose696 reviewed (commented):
Toolbar changes look good to me. Is there a particular advantage of the changes on the event handling presenters? I think registering too many listeners without necessarily needing an event handling mechanism might become an issue for larger projects

dapp commented:
> Toolbar changes look good to me. Is there a particular advantage of the changes on the event handling presenters? I think registering too many listeners without necessarily needing an event handling mechanism might become an issue for larger projects

It just seemed like event handling is frequently needed, more-so than not.   I don't think there should be too many listeners, since generally there is only one screen presenter registered for events at any time.  However, I do want to tinker with it a bit more, because I don't like forcing screen presenters to override the component event manager by default.  I might just remove those changes and push just the toolbar stuff.  stay tuned.

fkose696 commented on architecture/src/main/java/com/xfinity/blueprint/architecture/DefaultScreenView.kt:
```diff
@@ -45,6 +46,18 @@ open class ToolbarScreenView(screenViewDelegate: ScreenViewDelegate, messageView
     }
+
+    override fun setShowHomeAsUp(showHomeAsUp: Boolean) {
+        actionBar?.setDisplayHomeAsUpEnabled(showHomeAsUp)
```
Can we keep using show/hide functions for this?

dapp commented:
@fkose696  - I pushed quite a few new and breaking changes which I see as a major version bump.  This started out as just and expansion of toolbar support, but it's turned into more than that.

In the new changes, I revised some of the more rigid inheritance structures to make it easier for clients to use custom layouts for blueprint activities and fragments.

I've included a simple example in the sample app that show how to use the new system to add a FAB to an activity or fragment.  Since the FAB lays on top of everything else (not in the list) it requires a custom layout.  Before it was pretty difficult to use a custom layout, and to create a view and presenter to control it. 

Now, I think, it's way easier.  The number of classes and subclasses has also been greatly reduced, but with no loss in functionality.

fkose696 reviewed (dismissed):
Looks good to me

dapp commented:
@fkose696  -- since these are breaking changes, I added some revisions to the component view base classes.  This new change removes the set<>Image functions, and renames the existing one.

The reason I renamed the functions is because people love to put the word "image" in their image variable names, which yields generated function names like setImageImage().  

Instead, now we'll have setImageDrawable and setImageResource.

Let me know what you think

fkose696 reviewed (approved):
I agree. This should improve readability in the presenters
